### PR TITLE
feat: [ACI-854] implement penalty applying

### PR DESCRIPTION
**Pull request by andrii-hantkovskyi**

### CODE REVIEW

File: credentials/apps/badges/models.py [hunk 1]
```diff
@@ -265,6 +267,20 @@ def __str__(self):
 
     class Meta:
         verbose_name_plural = "Badge penalties"
+    
+    def apply_rules(self, data: dict) -> bool:
```
Comment (wowkalucky) [inline credentials/apps/badges/models.py:271]: @andrii-hantkovskyi can we get something like this:

```python
def apply_rules(self, data: dict) -> bool:
    return all(rule.apply(data) for rule in self.penaltydatarule_set.all())
```

Comment (wowkalucky) [reply credentials/apps/badges/models.py:271]: the same for requirements.

Comment (wowkalucky) [reply credentials/apps/badges/models.py:271]: let's also prettify
```
self.penaltydatarule_set.all() >> self.rules.all()
```

**Review by wowkalucky — CHANGES_REQUESTED** [review]

File: credentials/apps/badges/models.py (hunk 1 above)
Comment (andrii-hantkovskyi) [reply credentials/apps/badges/models.py:288]: done ^^

File: credentials/apps/badges/services/processing.py [hunk 2]
```diff
@@ -53,6 +53,17 @@ def identify_user(*, event_type, event_payload):
     """
     Identifies event user based on provided keyword arguments and returns the username.
 
+    if (
```
Comment (wowkalucky) [inline credentials/apps/badges/services/processing.py:56]: This was moved to `credentials/apps/badges/services/revocation.py#process_penalties`.

File: credentials/apps/badges/services/revocation.py [hunk 3]
```diff
@@ -39,23 +36,18 @@ def process_penalties(event_type, username, payload_dict):
             - emit BADGE_PROGRESS_INCOMPLETE >> handle_badge_incompletion
     """
 
-    # TEMP: remove this stub after processing is implemented
```
Comment (wowkalucky) [inline credentials/apps/badges/services/revocation.py:42]: Also a regression.

File: credentials/apps/badges/services/revocation.py [hunk 4]
```diff
@@ -39,23 +36,18 @@ def process_penalties(event_type, username, payload_dict):
...
-    # actual processing goes here:
-
-    # for penalty in penalties:
-    #     penalty.apply_rules(**kwargs)
-    #     penalty.assign(username)
+    for penalty in penalties:
+        if not penalty.is_active:
+            continue
+        if penalty.apply_rules(payload_dict):
+            [
```
Comment (wowkalucky) [inline credentials/apps/badges/services/revocation.py:45]: I'm looking for this:
```python
penalty.requirements.reset(username)

# or

for requirement in penalty.requirements.all():
    requirement.reset(username)
```

Comment (wowkalucky) [reply credentials/apps/badges/services/revocation.py:45]: e.g. a penalty can be set to affect not all badge progress (only some requirements)

**Review by wowkalucky — CHANGES_REQUESTED** [review]

**Review by wowkalucky — COMMENTED** [review]
Please see comments

**Review by wowkalucky — APPROVED** [review]

**Pull request merged**